Add Kubernetes 1.31 release blog (#47281)
kubernetes/website: content/en/blog/_posts/2024-08-14-Kubernetes-v1-31-Release.md

Reviewer: Filename should be `content/en/blog/_posts/2024-08-13-Kubernetes-v1-31-Release.md`

Author: updated!

---
layout: blog
title: 'Kubernetes v1.31: TITLE'
date: 2024-08-14
slug: kubernetes-v1-31-release
author: >
  [Kubernetes Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)
---
**Authors:** [Kubernetes 1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release_team.md)

## Release theme and logo
<Logo image size is recommended to be no more than 2160px/>

## Highlights of features graduating to Stable

_This is a selection of some of the improvements that are now stable following the v1.31 release._

### Improved ingress connectivity reliability for Kube-proxy

> Kube-proxy improved ingress connectivity reliability is GA in v1.31. One of the common problems with Load Balancers in Kubernetes is the synchronization between the different components involved to avoid traffic drop, this feature implements a mechanism in kube-proxy for load balancers to do connection draining for terminating Nodes and establish some best practices for cloud providers and Kubernetes load balancers implementations.

Reviewer: ```suggestion
Kube-proxy improved ingress connectivity reliability is GA in v1.31. One of the common problems with load balancers in Kubernetes is the synchronization between the different components involved to avoid traffic drop, this feature implements a mechanism in kube-proxy for load balancers to do connection draining for terminating Nodes and establish some best practices for cloud providers and Kubernetes load balancers implementations.
```

> exerpt from SIG:

Reviewer: ```suggestion
```

This work was done as part of [KEP #3836](https://github.com/kubernetes/enhancements/issues/3836) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network). /docs/reference/networking/virtual-ips/#external-traffic-policy

### AppAprmor support is now stable

Kubernetes support for AppArmor is now GA. Protect your Containers using AppArmor by setting the `appArmorProfile.type` field in the Container's `securityContext`. Note that before Kubernetes v1.30, AppArmor was controlled via annotations; starting in v1.30 it is contorlled using fields. Its recommended that you should migrate away from using annotations and start using the `appArmorProfile.type` field.

> To learn more read the [AppArmor tutorial](/docs/tutorials/security/apparmor/). This work was done as a part of [KEP #24](https://github.com/kubernetes/enhancements/issues/24) lead by  [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

Reviewer: Do we have any other docs for AppArmor yet? Ideally we would.

Author: as far as i can tell these are the docs we have. There is a [PR](https://github.com/kubernetes/website/pull/47254) for adding some more apparmor related docs, but its not merged yet.

Author: @sftim following up here,  dont think there are any other apparmor docs that i can find.

Reviewer: We could nudge release docs to nudge the SIG about adding some, but it's not directly relevant to this announcement.

    

### Persistent Volume last phase transition time

Persistent Volume last phase transition time feature moved to GA in v1.31. This feature adds a `PersistentVolumeStatus` field which holds a timestamp of when a PersistentVolume last transitioned to a different phase.
With this feature enabled, every PersistentVolume object will have a new field `.status.lastTransitionTime`, that holds a timestamp of
when the volume last transitioned its phase. This change is not immediate; the new field will be populated whenever a PersistentVolume is updated and first transitions between phases (`Pending`, `Bound`, or `Released`) after upgrading to Kubernetes v1.31.
This allows you to measure time between when a PersistentVolume moves from `Pending` to `Bound`. This can be also useful for providing metrics and SLOs.

> For more details about this feature please visit the [PersistentVolume documentation page](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/persistent-volume-v1/#PersistentVolumeStatus).

Reviewer: I'd link to `/docs/concepts/storage/persistent-volumes/`.

> For more details about this feature please visit the [PersistentVolume documentation page](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/persistent-volume-v1/#PersistentVolumeStatus).

Reviewer: (in the right hand nav of https://kubernetes.io/docs/concepts/storage/persistent-volumes/, the first link is to the API reference for PersistentVolume)

> For more details about this feature please visit the [PersistentVolume documentation page](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/persistent-volume-v1/#PersistentVolumeStatus).

Author: i've updated this link to /docs/concepts/storage/persistent-volumes/

In this mode, kube-proxy configures packet forwarding rules using the nftables API of the kernel netfilter subsystem. 
For each endpoint, it installs nftables rules which, by default, select a backend Pod at random.

The nftables API is the successor to the iptables API and is designed to provide better performance and scalability than iptables. 
The `nftables` proxy mode is able to process changes to service endpoints faster and more efficiently than the `iptables` mode, and is also able to more efficiently process packets in the kernel (though this only
becomes noticeable in clusters with tens of thousands of services).

As of Kubernetes v1.31, the `nftables` mode is still relatively new, and may not be compatible with all network plugins; consult the documentation for your network plugin. 
This proxy mode is only available on Linux nodes, and requires kernel 5.13 or later.
See the full [migration guide](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) for more details. 

This work was done as part of [KEP #3866](https://github.com/kubernetes/enhancements/issues/3866) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

> ### Always Honor PersistentVolume Reclaim Policy

Reviewer: ```suggestion
### Always honor PersistentVolume reclaim policy
```

or
```suggestion
### Changes to reclaim policy for PersistentVolumes
```

The Always Honor PersistentVolume Reclaim Policy feature has advanced to beta in Kubernetes v1.31. This enhancement ensures that the PersistentVolume (PV) reclaim policy is respected even after the associated PersistentVolumeClaim (PVC) is deleted, thereby preventing the leakage of volumes.

Prior to this feature, the reclaim policy linked to a PV could be disregarded under specific conditions, depending on whether the PV or PVC was deleted first. Consequently, the corresponding storage resource in the external infrastructure might not be removed, even if the reclaim policy was set to "Delete". This led to potential inconsistencies and resource leaks.

With the introduction of this feature, Kubernetes now guarantees that the "Delete" reclaim policy will be enforced, ensuring the deletion of the underlying storage object from the backend infrastructure, regardless of the deletion sequence of the PV and PVC.

This work was done as a part of [KEP #2644](https://github.com/kubernetes/enhancements/issues/2644) and lead by [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).
    

> ### Bound service account token improvements

Author: waiting for content, https://kubernetes.slack.com/archives/C0EN96KUY/p1722015619684019

Author: @enj and @ritazh moved content into the PR. @enj had concerns about this content based on the slack thread. Can you help suggest some way to improve the wording?

Author: moved to other comment thread, https://github.com/kubernetes/website/pull/47281/files#r1711543988

IP exhaustion has been a problem for long lived or large clusters, as admins needed to expand, shrink or even replace entirely the assigned Service CIDR range. 
These operations were never supported natively and were performed via complex and delicate maintenance operations, often causing downtime on their clusters. This KEP allows users and cluster admins to dynamically modify Service CIDR ranges with zero downtime.

For more details about this feature please visit the [Virtual IPs and Service Proxies documentation page](https://kubernetes.io/docs/reference/networking/virtual-ips/#ip-address-objects)

This work was done as part of [KEP #1880](https://github.com/kubernetes/enhancements/issues/1880) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

### Traffic Distribution for Services

Traffic Distribution for Services moves to beta in v1.31, enabled by default. 

> After several iterations on finding the best user experience and traffic engineering capabilities for Services networking, the SIG Networking implemented a `trafficDistribution` field in the Service specification, that serves as a guideline for the underlying implementation to consider while making routing decisions.

Reviewer 2: ```suggestion
After several iterations on finding the best user experience and traffic engineering capabilities for Services networking, the SIG Networking implemented the `trafficDistribution` field in the Service specification, that serves as a guideline for the underlying implementation to consider while making routing decisions.
```

One of the requirements to support these use cases is to support Open Container Initiative (OCI) compatible images and artifacts (referred as OCI objects) directly as a native volume source. This allows users to focus on OCI standards as well as enables them to store and distribute any content using OCI registries.

Given that, the Image Volume Source feature is introduced as a new alpha feature in v1.31. This feature allows users to specify an image reference as volume in a pod while reusing it as volume mount within containers.

This work was done as part of [KEP #4639](https://github.com/kubernetes/enhancements/issues/4639) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) and [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Better devices errors visibility - the device health is now in pod status: Add Resource Health Status to the Pod Status for Device Plugin and DRA

Expose device health information through the Pod Status is added as a new alpha feature in v1.31, disabled by default.

> Before Kubernetes v1.31, the way to know whether or not a Pod is associated with the failed device is to use the [PodResources API](#monitoring-device-plugin-resources). 

Reviewer 2: ```suggestion
Before Kubernetes v1.31, the way to know whether or not a Pod is associated with the failed device is to use the [PodResources API](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#monitoring-device-plugin-resources). 
```

This release includes a total of 11 enhancements promoted to Stable:

* [PersistentVolume last phase transition time](	https://github.com/kubernetes/enhancements/issues/3762)
* [Metric cardinality enforcement](	https://github.com/kubernetes/enhancements/issues/2305)
* [Kube-proxy improved ingress connectivity reliability](	https://github.com/kubernetes/enhancements/issues/3836)
* [Add CDI devices to device plugin API](	https://github.com/kubernetes/enhancements/issues/4009)
* [Move cgroup v1 support into maintenance mode](	https://github.com/kubernetes/enhancements/issues/4569)
* [AppArmor support](	https://github.com/kubernetes/enhancements/issues/24)
* [PodHealthyPolicy for PodDisruptionBudget](	https://github.com/kubernetes/enhancements/issues/3017)
* [Retriable and non-retriable Pod failures for Jobs](	https://github.com/kubernetes/enhancements/issues/3329)
* [Elastic Indexed Jobs](	https://github.com/kubernetes/enhancements/issues/3715)
* [Allow StatefulSet to control start replica ordinal numbering](	https://github.com/kubernetes/enhancements/issues/3335)
* [Random Pod selection on ReplicaSet downscaling](	https://github.com/kubernetes/enhancements/issues/2185)
    
    

> ### Deprecations and Removals 

Reviewer 3: Please add a new removal chapter with something like this (feel free to rephrase):

#### InTreePluginXXXUnregister feature gates removed.

Alpha feature gates `InTreePluginAWSUnregister`, `InTreePluginAzureDiskUnregister`, `InTreePluginAzureFileUnregister`, `InTreePluginGCEUnregister`, `InTreePluginOpenStackUnregister` and `InTreePluginvSphereUnregister` were removed in Kubernetes 1.31.

These feature gates allowed us to test situation when the in-tree volume plugins were removed from code without actually removing them. The mentioned volume plugins were removed already in Kubernetes 1.30 and these feature gates did nothing useful.

The only remaining CSI migration gate is `InTreePluginPortworxUnregister` and it remains alpha until the CSI migration for Portworx is complete and the in-tree volume plugin is ready for removal.

\<end of suggested content>

* PR that removed the featgure gates: https://github.com/kubernetes/kubernetes/pull/124815 
* Issue about missing documentation: https://github.com/kubernetes/kubernetes/issues/126406

Reviewer: +1 from me; this is important to explain.

Author: @sftim do you mean that we should add more explanation to this section? 

Do you think the link to the previous blog is enough? or should we pull in all the deprecations/removals from there into this blog as well?

Author: I've added in all the deprecations and removals from the mid-cycle blog.

This shift aligns with the broader industry's move towards [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), offering improved functionality, scalability, and a more consistent interface. 
Maintance mode means that no new features will be added to cgroup v1 support. 
Critical security fixes will still be provided, however, bug-fixing is now best-effort, meaning major bugs may be fixed if feasible, but some issues might remain unresolved.

Its recommended that you start switching to use cgroup v2 as soon as possible. For more details on the reasoning behind placing cgroup v1 in maintaince mode and starting the migration process see the [Moving cgroup v1 Support into Maintenance Mode in Kubernetes 1.31](/blog/kubernetes-1-31-moving-cgroup-v1-support-maintenance-mode) article. 

Please report any problems encounter by filing an [issue](https://github.com/kubernetes/kubernetes).

This work was done as part of [KEP #4569](https://github.com/kubernetes/enhancements/issues/4569) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

> 

Reviewer 3: Please add a new removal chapter with something like this (feel free to rephrase):

```suggestion

####  InTreePluginXXXUnregister feature gates removed.

Alpha feature gates `InTreePluginAWSUnregister`, `InTreePluginAzureDiskUnregister`, `InTreePluginAzureFileUnregister`, `InTreePluginGCEUnregister`, `InTreePluginOpenStackUnregister` and `InTreePluginvSphereUnregister` were removed in Kubernetes 1.31.

These feature gates allowed us to test situation when the in-tree volume plugins were removed from code without actually removing them. The mentioned volume plugins were removed already in Kubernetes 1.30 and these feature gates did nothing useful.

The only remaining CSI migration gate is `InTreePluginPortworxUnregister` and it remains alpha until the CSI migration for Portworx is complete and the in-tree volume plugin is ready for removal.

```

* PR that removed the feature gates: https://github.com/kubernetes/kubernetes/pull/124815 
* Issue about missing documentation: https://github.com/kubernetes/kubernetes/issues/126406

Reviewer 4: This is a possible version

```
In Kubernetes 1.31, the following alpha feature gates `InTreePluginAWSUnregister`, `InTreePluginAzureDiskUnregister`, `InTreePluginAzureFileUnregister`, `InTreePluginGCEUnregister`, `InTreePluginOpenStackUnregister`, and `InTreePluginvSphereUnregister` have been removed.

These feature gates were introduced to facilitate the testing of scenarios where in-tree volume plugins were removed from the codebase, without actually removing them. Since Kubernetes 1.30 had deprecated these in-tree volume plugins, these feature gates were redundant and no longer served a purpose.

The only CSI migration gate still standing is `InTreePluginPortworxUnregister`, which will remain in alpha until the CSI migration for Portworx is completed and its in-tree volume plugin will be ready for removal.
```

Author: @jsafrane I've added in this information, can you review

Reviewer 3: Looks good to me, thanks!

Author: thanks @jsafrane and @mbianchidev resolving this comment :)

> ## Release Notes and Upgrade Actions Required

Reviewer: ```suggestion
## Release notes and upgrade actions required
```

Check out the full details of the Kubernetes 1.31 release in our [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.31.md).

> ### KubeSchedule now uses QueueingHint when `SchedulerQueueingHints` is enabled

Reviewer: ```suggestion
### Scheduler now uses QueueingHint when `SchedulerQueueingHints` is enabled
```

> Added support to the scheduler to start using QueueingHint registered for Pod/Updated event to determine whether unschedulable Pods update make them schedulable, when the feature gate `SchedulerQueueingHints` is enabled. Previously, when unschedulable Pods are updated, the scheduler always put Pods back to activeQ/backoffQ. But, actually not all updates to Pods make Pods schedulable, especially considering many scheduling constraints nowadays are immutable. Now, when unschedulable Pods are updated, the scheduling queue checks with QueueingHint(s) whether the update may make the pods schedulable, and requeues them to activeQ/backoffQ only when at least one QueueingHint(s) return Queue.

Reviewer: ```suggestion
Added support to the scheduler to start using a QueueingHint registered for Pod/Updated events,
to determine whether updates to  previously unschedulable Pods have made them schedulable.
The new support is active when the feature gate `SchedulerQueueingHints` is enabled.

Previously, when unschedulable Pods were updated, the scheduler always put Pods back to into a queue
(`activeQ` / `backoffQ`). However not all updates to Pods make Pods schedulable, especially considering
many scheduling constraints nowadays are immutable. Under the new behaviour, once unschedulable Pods
are updated, the scheduling queue checks with QueueingHint(s) whether the update may make the
pod(s) schedulable, and requeues them to `activeQ` or `backoffQ` only when at least one
QueueingHint returns `Queue`.
```

**Action required for custom scheduler plugin developers**: 

> Plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods themselves. Example: suppose you develop a custom plugin that denies Pods that have a schedulable=false label. Given Pods with a schedulable=false label will be schedulable if the schedulable=false label is removed, this plugin would implement QueueingHint for Pod/Update event that returns Queue when such label changes are made in unscheduled Pods. [#122234](https://github.com/kubernetes/kubernetes/pull/122234)

Reviewer 2: ```suggestion
Plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods themselves. Example: suppose you develop a custom plugin that denies Pods that have a `schedulable=false` label. Given Pods with a `schedulable=false` label will be schedulable if the `schedulable=false` label is removed, this plugin would implement QueueingHint for Pod/Update event that returns Queue when such label changes are made in unscheduled Pods. [#122234](https://github.com/kubernetes/kubernetes/pull/122234)
```

### Removal of kubelet --keep-terminated-pod-volumes command line flag

> The kubelet flag --keep-terminated-pod-volumes, which was deprecated in 2017, will be removed as part of the v1.31 release.

Reviewer 2: ```suggestion
The kubelet flag `--keep-terminated-pod-volumes`, which was deprecated in 2017, will be removed as part of the v1.31 release.
```

Reviewer: ```suggestion
The kubelet flag `--keep-terminated-pod-volumes`, which was deprecated in 2017, was removed as part of the v1.31 release.
```

You can find more details in the pull request [#122082](https://github.com/kubernetes/kubernetes/pull/122082).

## Availability

Kubernetes 1.31 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.31.0). To get started with Kubernetes, check out these [interactive tutorials](https://kubernetes.io/docs/tutorials/) or run local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily install 1.31 using [kubeadm](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/). 

> ## Release Team

Reviewer: ```suggestion
## Release team
```

Kubernetes is only possible with the support, commitment, and hard work of its community. Each release team is made up of dedicated community volunteers who work together to build the many pieces that make up the Kubernetes releases you rely on. This requires the specialized skills of people from all corners of our community, from the code itself to its documentation and project management.

We would like to thank the entire [release team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md) for the hours spent hard at work to deliver the Kubernetes v1.31 release to our community. The Release Team's membership ranges from first-time shadows to returning team leads with experience forged over several release cycles. A very special thanks goes out our release lead, Angelos Kolaitis, for supporting us through a successful release cycle, advocating for us, making sure that we could all contribute in the best way possible, and challenging us to improve the release process.

> ## Project Velocity

Reviewer: ```suggestion
## Project velocity
```

Source for this data: 
- [Companies contributing to Kubernetes](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1715032800000&to=1723586399000&var-period=d28&var-repogroup_name=Kubernetes&var-repo_name=kubernetes%2Fkubernetes)
- [Overall ecosystem contributions](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&from=1715032800000&to=1723586399000&var-period=d28&var-repogroup_name=All&var-repo_name=kubernetes%2Fkubernetes)

By contribution we mean when someone makes a commit, code review, comment, creates an issue or PR, reviews a PR (including blogs and documentation) or comments on issues and PRs.

If you are interested in contributing visit [this page](https://www.kubernetes.dev/docs/guide/#getting-started) to get started. 

[Check out DevStats](https://k8s.devstats.cncf.io/d/11/companies-contributing-in-repository-groups?orgId=1&var-period=m&var-repogroup_name=All) to learn more about the overall velocity of the Kubernetes project and community.

> ## Event Update

Reviewer: ```suggestion
## Event update
```

Explore the upcoming Kubernetes and cloud-native events from August to November 2024, featuring KubeCon, KCD, and other notable conferences worldwide. Stay informed and engage with the Kubernetes community.

**August 2024**

> - [**KCD Taipei 2024**](https://community.cncf.io/events/details/cncf-kcd-taiwan-presents-kcd-taipei-2024/): August 3-4, 2024 | Taipei City, Taiwan

Reviewer: This will have happened.

- [**KCD Porto 2024**](https://community.cncf.io/events/details/cncf-kcd-porto-presents-kcd-porto-2024/): September 27-28, 2024 | Porto, Portugal

**October 2024**
- [**KubeDay Australia**](https://events.linuxfoundation.org/kubeday-australia/): October 1, 2024 | Melbourne, Australia
- [**KCD Austria 2024: October 8-10, 2024 | Wien, Austria**](https://community.cncf.io/events/details/cncf-kcd-austria-presents-kcd-austria-2024/) 
- [**KCD UK - London 2024**](https://community.cncf.io/events/details/cncf-kcd-uk-presents-kubernetes-community-days-uk-london-2024/): October 22-23, 2024 | Greater London, United Kingdom

**November 2024**
- [**KubeCon + CloudNativeCon North America 2024**](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/): November 12-15, 2024 | Salt Lake City, United States
- [**Kubernetes on EDGE Day North America**](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/co-located-events/kubernetes-on-edge-day/): November 12, 2024 | Salt Lake City, United States

> ## Upcoming Release Webinar

Reviewer: ## Upcoming release webinar

> <RELEASE WEBINARE WILL TAKE PLACE NORMALLY 30 DAYS AFTER RELEASE, ALIGN WITH CNCF TO HIGHLIGHT THE WEBINAR>

Reviewer: FIXME!

Author: I've added the webinar info, but its not on the cncf website yet. so need to verify still

> ## Get Involved

Reviewer: ```suggestion
## Get involved
```

---
layout: blog
title: 'Kubernetes v1.31: TITLE'
date: 2024-08-14
slug: kubernetes-v1-31-release
author: >
  [Kubernetes Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)
---
**Authors:** [Kubernetes 1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release_team.md)

> ## Release theme and logo

Author: @neoaggelos can you send this over when you have the theme and logo ready

* [Metric cardinality enforcement](	https://github.com/kubernetes/enhancements/issues/2305)
* [Kube-proxy improved ingress connectivity reliability](	https://github.com/kubernetes/enhancements/issues/3836)
* [Add CDI devices to device plugin API](	https://github.com/kubernetes/enhancements/issues/4009)
* [Move cgroup v1 support into maintenance mode](	https://github.com/kubernetes/enhancements/issues/4569)
* [AppArmor support](	https://github.com/kubernetes/enhancements/issues/24)
* [PodHealthyPolicy for PodDisruptionBudget](	https://github.com/kubernetes/enhancements/issues/3017)
* [Retriable and non-retriable Pod failures for Jobs](	https://github.com/kubernetes/enhancements/issues/3329)
* [Elastic Indexed Jobs](	https://github.com/kubernetes/enhancements/issues/3715)
* [Allow StatefulSet to control start replica ordinal numbering](	https://github.com/kubernetes/enhancements/issues/3335)
* [Random Pod selection on ReplicaSet downscaling](	https://github.com/kubernetes/enhancements/issues/2185)
    
    
### Deprecations and Removals 

Also see our [Kubernetes Removals and Major Changes In v1.31](https://kubernetes.io/blog/2024/07/19/kubernetes-1-31-upcoming-changes/) blog.
    

> #### Cgroup v1 enters the maintenance mode

Author: There is a blog going out https://github.com/kubernetes/website/pull/47069, currently set for 8/15. Should we move the date of that blog to go out the same day as the release? I'd like to make sure we don't create confusion about this topic so i think it would be good if we can link to this blog to give readers more context

Author: the content as written, currently links to the blog. So we'd have to update that if we dont publish the blogs the same day.

Author: @harche can you review this content around the cgroup maintenance mode for the release blog?

Reviewer 5: @rashansmith  what do you think about the publishing date?

xref: https://github.com/kubernetes/enhancements/issues/4569#issuecomment-2182864829

Author: @harche, yes, lets move up the publication date to August 13th your [blog](https://github.com/kubernetes/website/pull/47069). It will go out the same day as the release so we can link readers to your blog for further reading.

Author: please also see https://github.com/kubernetes/website/pull/47069#issuecomment-2277953863. I've updated the content here to include more information in the release blog and no longer link to the https://github.com/kubernetes/website/pull/47069 post. This way the https://github.com/kubernetes/website/pull/47069 can go out on august 14th and we dont have to worry about timestamps. This release blog will be the only blog we publish August 13th and feature blogs will start publishing august 14th.

Reviewer: We can _if we want_ change #47069, post release. The change I'm imagining would revise this article with a link to the new post.

This feature enables connection draining for workloads running on terminating/deleting Kubernetes nodes which are exposed by services of `type: LoadBalancer` and `externalTrafficPolicy: Cluster`. Note: the predicates for using this feature are: a) kube-proxy needs to run as default service proxy on the cluster and b) the load balancer needs to support connection draining. There are no specific changes required for using this feature, it has been enabled by default in Kube-proxy since 1.30 and been promoted to stable in 1.31.

For more details about this feature please visit the [Virtual IPs and Service Proxies documentation page](/docs/reference/networking/virtual-ips/#external-traffic-policy).

This work was done as part of [KEP #3836](https://github.com/kubernetes/enhancements/issues/3836) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

> ### AppAprmor support is now stable

Reviewer: (nit / opinion)

I'd mention this one first.

Author: i moved the apparmor section to be first

With this feature enabled, every PersistentVolume object will have a new field `.status.lastTransitionTime`, that holds a timestamp of
when the volume last transitioned its phase. This change is not immediate; the new field will be populated whenever a PersistentVolume is updated and first transitions between phases (`Pending`, `Bound`, or `Released`) after upgrading to Kubernetes v1.31.
This allows you to measure time between when a PersistentVolume moves from `Pending` to `Bound`. This can be also useful for providing metrics and SLOs.

For more details about this feature please visit the [PersistentVolume documentation page](https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/persistent-volume-v1/#PersistentVolumeStatus).

This work was done as a part of [KEP #3762](https://github.com/kubernetes/enhancements/issues/3762) lead by  [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

## Highlights of features graduating to Beta

_This is a selection of some of the improvements that are now beta following the v1.31 release._
    

> ### Kube-proxy nftables backend
> 
> Kube-proxy nftables backend moves to beta in v1.31 and introduces a new feature gate, `NFTablesProxyMode`, disabled by default.

Reviewer: To avoid confusion about sentence case and component names, we could write:
```suggestion
### nftables backend for kube-proxy

The Kubernetes project intends to update kube-proxy on Linux so that it uses `nftables`,
not `iptables`, as the default implementation. The nftables backend moves to beta in v1.31,
behind a newly added feature gate, `NFTablesProxyMode` (disabled by default).
```

Prior to this feature, the reclaim policy linked to a PV could be disregarded under specific conditions, depending on whether the PV or PVC was deleted first. Consequently, the corresponding storage resource in the external infrastructure might not be removed, even if the reclaim policy was set to "Delete". This led to potential inconsistencies and resource leaks.

With the introduction of this feature, Kubernetes now guarantees that the "Delete" reclaim policy will be enforced, ensuring the deletion of the underlying storage object from the backend infrastructure, regardless of the deletion sequence of the PV and PVC.

This work was done as a part of [KEP #2644](https://github.com/kubernetes/enhancements/issues/2644) and lead by [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).
    
### Bound service account token improvements

This work was done as part of [KEP #4193](https://github.com/kubernetes/enhancements/issues/4193) by [SIG Auth](https://github.com/kubernetes/community/tree/master/sig-auth).

### Multiple Service CIDRs

> Multiple Service CIDRs moves to beta in v1.31 (disabled by default).

Reviewer: ```suggestion
Support for clusters with multiple Service CIDRs moves to beta in v1.31 (disabled by default).
```
?

For more details about this feature please visit the [Virtual IPs and Service Proxies documentation page](https://kubernetes.io/docs/reference/networking/virtual-ips/#ip-address-objects).

This work was done as part of [KEP #1880](https://github.com/kubernetes/enhancements/issues/1880) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

### Traffic distribution for Services

Traffic Distribution for Services moves to beta in v1.31, enabled by default. 

After several iterations on finding the best user experience and traffic engineering capabilities for Services networking, the SIG Networking implemented the `trafficDistribution` field in the Service specification, that serves as a guideline for the underlying implementation to consider while making routing decisions.
It supersedes the functionality formerly provided by the `service.kubernetes.io/topology-mode` annotation and its precursor `topologyKeys` field (which has been deprecated since Kubernetes 1.21)

> For more details about this feature please read the [1.30 Release Blog](https://kubernetes.io/blog/2024/04/17/kubernetes-v1-30-release/#traffic-distribution-for-services-sig-network-https-github-com-kubernetes-community-tree-master-sig-network) or visit the [Service documentation page](https://kubernetes.io/docs/concepts/services-networking/service/#traffic-distribution).

Reviewer: ```suggestion
For more details about this feature please read the
[1.30 Release Blog](/blog/2024/04/17/kubernetes-v1-30-release/#traffic-distribution-for-services-sig-network-https-github-com-kubernetes-community-tree-master-sig-network)
or visit the [Service](/docs/concepts/services-networking/service/#traffic-distribution) documentation page.
```

This work was done as part of [KEP #4444](https://github.com/kubernetes/enhancements/issues/4444) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

### Kubernetes VolumeAttributesClass ModifyVolume

[VolumeAttributesClass](/docs/concepts/storage/volume-attributes-classes/) API is moving to beta in v1.31.
The VolumeAttributesClass provides a generic,

> Kubernetes-native API for modifying dynamically volume parameters like provisioned IO. This allows workloads to vertically scale their volumes on-line to balance cost and performance, if supported by their provider. This feature has been alpha since 1.29 and will be beta in 1.31. 

Reviewer: ```suggestion
Kubernetes-native API for modifying dynamically volume parameters like provisioned IO.
This allows workloads to vertically scale their volumes on-line to balance cost and performance, if supported by their provider.
This feature had been alpha since Kubernetes 1.29.
```
The tense is wrong for a post-release announcement; the suggestion aims to fix that.

This work was done as part of [KEP #4639](https://github.com/kubernetes/enhancements/issues/4639) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) and [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

### Exposing device health information through Pod status

Expose device health information through the Pod Status is added as a new alpha feature in v1.31, disabled by default.

Before Kubernetes v1.31, the way to know whether or not a Pod is associated with the failed device is to use the [PodResources API](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#monitoring-device-plugin-resources). 

By enabling this feature, the field `allocatedResourcesStatus` will be added to each container status, within the `.status` for each Pod. The `allocatedResourcesStatus` field reports health information for each device assigned to the container.
 

This work was done as part of [KEP #4680](https://github.com/kubernetes/enhancements/issues/4680) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

### Finer-grained authorization based on selectors

> This feature allows webhook authorizers and future (but not currently designed) in-tree authorizers to allow list and watch requests based on field and label selectors.  It is now possible for an authorizer to express: this user cannot list all pods, but can list all pods where `.spec.nodeName=foo`.  Combined with CRD field selectors (moving to beta in 1.31), it is possible to write more secure per-node extensions.

Reviewer: (nit)
I like to avoid pseudocode in articles as it can encourage readers to build a misleading mental model of
how Kubernetes works.

Do we mention CRD field selectors in the article? If not, I think we should.
```suggestion
This feature allows webhook authorizers and future (but not currently designed) in-tree authorizers to
allow **list** and **watch** requests, provided those requests use label and/or field selectors.
For example, it is now possible for an authorizer to express: this user cannot list all pods, but can list all pods where `.spec.nodeName` matches some specific value. Or to allow a user to watch all Secrets in a namespace
that are _not_ labelled as `confidential: true`.
Combined with CRD field selectors (also moving to beta in 1.31), it is possible to write more secure
per-node extensions.
```

---
layout: blog
title: 'Kubernetes v1.31: TITLE'
date: 2024-08-14
slug: kubernetes-v1-31-release
author: >

>   [Kubernetes Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)

Reviewer 6: ```suggestion
  [Kubernetes v1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)
```

---

> **Authors:** [Kubernetes 1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release_team.md)

Reviewer 6: This 'Authors' section can be removed from the content now that we've included it in the front matter of the blog file.
```suggestion
```

With this feature enabled, every PersistentVolume object will have a new field `.status.lastTransitionTime`, that holds a timestamp of
when the volume last transitioned its phase. This change is not immediate; the new field will be populated whenever a PersistentVolume is updated and first transitions between phases (`Pending`, `Bound`, or `Released`) after upgrading to Kubernetes v1.31.
This allows you to measure time between when a PersistentVolume moves from `Pending` to `Bound`. This can be also useful for providing metrics and SLOs.

For more details about this feature please visit the [PersistentVolume documentation page](/docs/concepts/storage/persistent-volumes/).

This work was done as a part of [KEP #3762](https://github.com/kubernetes/enhancements/issues/3762) lead by  [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

## Highlights of features graduating to Beta

_This is a selection of some of the improvements that are now beta following the v1.31 release._
    

> ### nftables backend for kube-proxy

Author: @danwinship can you review this section of the release blog?

Author: resolving this for a new comment, https://github.com/kubernetes/website/pull/47281/files#r1711552290

* [PersistentVolume last phase transition time](	https://github.com/kubernetes/enhancements/issues/3762)
* [Metric cardinality enforcement](	https://github.com/kubernetes/enhancements/issues/2305)
* [Kube-proxy improved ingress connectivity reliability](	https://github.com/kubernetes/enhancements/issues/3836)
* [Add CDI devices to device plugin API](	https://github.com/kubernetes/enhancements/issues/4009)
* [Move cgroup v1 support into maintenance mode](	https://github.com/kubernetes/enhancements/issues/4569)
* [AppArmor support](	https://github.com/kubernetes/enhancements/issues/24)
* [PodHealthyPolicy for PodDisruptionBudget](	https://github.com/kubernetes/enhancements/issues/3017)
* [Retriable and non-retriable Pod failures for Jobs](	https://github.com/kubernetes/enhancements/issues/3329)
* [Elastic Indexed Jobs](	https://github.com/kubernetes/enhancements/issues/3715)
* [Allow StatefulSet to control start replica ordinal numbering](	https://github.com/kubernetes/enhancements/issues/3335)
* [Random Pod selection on ReplicaSet downscaling](	https://github.com/kubernetes/enhancements/issues/2185)
    
    
### Deprecations and Removals 

> Also see our [Kubernetes Removals and Major Changes In v1.31](https://kubernetes.io/blog/2024/07/19/kubernetes-1-31-upcoming-changes/) blog.

Reviewer 6: ```suggestion
Also see our [Kubernetes Removals and Major Changes In v1.31](/blog/2024/07/19/kubernetes-1-31-upcoming-changes/) blog.
```

are updated, the scheduling queue checks with QueueingHint(s) whether the update may make the
pod(s) schedulable, and requeues them to `activeQ` or `backoffQ` only when at least one
QueueingHint returns `Queue`.

**Action required for custom scheduler plugin developers**: 
Plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods themselves. Example: suppose you develop a custom plugin that denies Pods that have a `schedulable=false` label. Given Pods with a `schedulable=false` label will be schedulable if the `schedulable=false` label is removed, this plugin would implement QueueingHint for Pod/Update event that returns Queue when such label changes are made in unscheduled Pods. [#122234](https://github.com/kubernetes/kubernetes/pull/122234)

### Removal of kubelet --keep-terminated-pod-volumes command line flag
The kubelet flag `--keep-terminated-pod-volumes`, which was deprecated in 2017, was removed as part of the v1.31 release.

You can find more details in the pull request [#122082](https://github.com/kubernetes/kubernetes/pull/122082).

## Availability

> 

Reviewer: BTW, we could also mention that https://kubernetes.io/releases/download/ exists.

Reviewer 4: Here's the fixed version

```
Kubernetes 1.31 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.31.0) or on the [Kubernetes download page](https://kubernetes.io/releases/download/). 

To get started with Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily install 1.31 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/). 
```

Author: updated.

> Kubernetes 1.31 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.31.0). To get started with Kubernetes, check out these [interactive tutorials](https://kubernetes.io/docs/tutorials/) or run local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily install 1.31 using [kubeadm](https://kubernetes.io/docs/setup/independent/create-cluster-kubeadm/). 

Reviewer 6: ```suggestion
Kubernetes 1.31 is available for download on [GitHub](https://github.com/kubernetes/kubernetes/releases/tag/v1.31.0). To get started with Kubernetes, check out these [interactive tutorials](/docs/tutorials/) or run local Kubernetes clusters using [minikube](https://minikube.sigs.k8s.io/). You can also easily install 1.31 using [kubeadm](/docs/setup/independent/create-cluster-kubeadm/). 
```

---
layout: blog
title: 'Kubernetes v1.31: TITLE'
date: 2024-08-14
slug: kubernetes-v1-31-release
author: >
  [Kubernetes v1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)
---

## Release theme and logo
<Logo image size is recommended to be no more than 2160px/>

## Highlights of features graduating to Stable

_This is a selection of some of the improvements that are now stable following the v1.31 release._

### AppAprmor support is now stable

> Kubernetes support for AppArmor is now GA. Protect your Containers using AppArmor by setting the `appArmorProfile.type` field in the Container's `securityContext`. Note that before Kubernetes v1.30, AppArmor was controlled via annotations; starting in v1.30 it is controlled using fields. It is recommended that you should migrate away from using annotations and start using the `appArmorProfile.type` field.

Reviewer 2: ```suggestion
Kubernetes support for AppArmor is now GA. Protect your containers using AppArmor by setting the `appArmorProfile.type` field in the container's `securityContext`. Note that before Kubernetes v1.30, AppArmor was controlled via annotations; starting in v1.30 it is controlled using fields. It is recommended that you should migrate away from using annotations and start using the `appArmorProfile.type` field.
```

> To learn more read the [AppArmor tutorial](/docs/tutorials/security/apparmor/). This work was done as a part of [KEP #24](https://github.com/kubernetes/enhancements/issues/24) lead by  [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

Reviewer 7: nit: extra space
```suggestion
To learn more read the [AppArmor tutorial](/docs/tutorials/security/apparmor/). This work was done as a part of [KEP #24](https://github.com/kubernetes/enhancements/issues/24) lead by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).
```

Author: i believe this has been addressed in another commit, thanks!

> To learn more read the [AppArmor tutorial](/docs/tutorials/security/apparmor/). This work was done as a part of [KEP #24](https://github.com/kubernetes/enhancements/issues/24) lead by  [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).

Reviewer: I recommend one extra tweak (“led“):
```suggestion
To learn more read the [AppArmor tutorial](/docs/tutorials/security/apparmor/). This work was done as a part of [KEP #24](https://github.com/kubernetes/enhancements/issues/24), led by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).
```
We can also use the short link https://kep.k8s.io/24

> ### Improved ingress connectivity reliability for Kube-proxy

Reviewer 2: According to the [documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/), I guess it should start with a lowercase letter?
```suggestion
### Improved ingress connectivity reliability for kube-proxy
```

Kube-proxy improved ingress connectivity reliability is stable in v1.31. One of the common problems with load balancers in Kubernetes is the synchronization between the different components involved to avoid traffic drop. This feature implements a mechanism in kube-proxy for load balancers to do connection draining for terminating Nodes exposed by services of `type: LoadBalancer` and `externalTrafficPolicy: Cluster` and establish some best practices for cloud providers and Kubernetes load balancers implementations.

> To use this feature, kube-proxy needs to run as default service proxy on the cluster and the load balancer needs to support connection draining. There are no specific changes required for using this feature, it has been enabled by default in Kube-proxy since 1.30 and been promoted to stable in 1.31.

Reviewer 2: ```suggestion
To use this feature, kube-proxy needs to run as default service proxy on the cluster and the load balancer needs to support connection draining. There are no specific changes required for using this feature, it has been enabled by default in kube-proxy since 1.30 and been promoted to stable in 1.31.
```

This work was done as part of [KEP #4193](https://github.com/kubernetes/enhancements/issues/4193) by [SIG Auth](https://github.com/kubernetes/community/tree/master/sig-auth).

### Multiple Service CIDRs

Support for clusters with multiple Service CIDRs moves to beta in v1.31 (disabled by default).

There are multiple components in a Kubernetes cluster that consume IP addresses: Nodes, Pods and Services. Nodes and Pods IP ranges can be dynamically changed because depend on the infrastructure or the network plugin respectively. However, Services are defined during the cluster creation as a hardcoded flag in the kube-apiserver. 
IP exhaustion has been a problem for long lived or large clusters, as admins needed to expand, shrink or even replace entirely the assigned Service CIDR range. 

> These operations were never supported natively and were performed via complex and delicate maintenance operations, often causing downtime on their clusters. This KEP allows users and cluster admins to dynamically modify Service CIDR ranges with zero downtime.

Reviewer 2: ```suggestion
These operations were never supported natively and were performed via complex and delicate maintenance operations, often causing downtime on their clusters. This new feature allows users and cluster admins to dynamically modify Service CIDR ranges with zero downtime.
```

For more details about this feature please visit the
[Virtual IPs and Service Proxies](/docs/reference/networking/virtual-ips/#ip-address-objects) documentation page.

This work was done as part of [KEP #1880](https://github.com/kubernetes/enhancements/issues/1880) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

### Traffic distribution for Services

> Traffic Distribution for Services moves to beta in v1.31, enabled by default. 

Reviewer 7: nit: Does this improve readability? To me it does but I can see that it does not matter to most
```suggestion
Traffic distribution for Services moves to beta in v1.31 and is enabled by default. 
```

After several iterations on finding the best user experience and traffic engineering capabilities for Services networking, the SIG Networking implemented the `trafficDistribution` field in the Service specification, that serves as a guideline for the underlying implementation to consider while making routing decisions.

> It supersedes the functionality formerly provided by the `service.kubernetes.io/topology-mode` annotation and its precursor `topologyKeys` field (which has been deprecated since Kubernetes 1.21)

Reviewer 2: ```suggestion
It supersedes the functionality formerly provided by the `service.kubernetes.io/topology-mode` annotation and its precursor `topologyKeys` field (which has been deprecated since Kubernetes 1.21).
```

* [Add CDI devices to device plugin API](	https://github.com/kubernetes/enhancements/issues/4009)
* [Move cgroup v1 support into maintenance mode](	https://github.com/kubernetes/enhancements/issues/4569)
* [AppArmor support](	https://github.com/kubernetes/enhancements/issues/24)
* [PodHealthyPolicy for PodDisruptionBudget](	https://github.com/kubernetes/enhancements/issues/3017)
* [Retriable and non-retriable Pod failures for Jobs](	https://github.com/kubernetes/enhancements/issues/3329)
* [Elastic Indexed Jobs](	https://github.com/kubernetes/enhancements/issues/3715)
* [Allow StatefulSet to control start replica ordinal numbering](	https://github.com/kubernetes/enhancements/issues/3335)
* [Random Pod selection on ReplicaSet downscaling](	https://github.com/kubernetes/enhancements/issues/2185)
    
    
### Deprecations and Removals 

Also see our [Kubernetes Removals and Major Changes In v1.31](/blog/2024/07/19/kubernetes-1-31-upcoming-changes/) blog.
    
#### Cgroup v1 enters the maintenance mode

As Kubernetes continues to evolve and adapt to the changing landscape of container orchestration, the community has decided to move cgroup v1 support into maintenance mode in v1.31.

> This shift aligns with the broader industry's move towards [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), offering improved functionality, scalability, and a more consistent interface. 

Reviewer 2: ```suggestion
This shift aligns with the broader industry's move towards [cgroup v2](/docs/concepts/architecture/cgroups/), offering improved functionality, scalability, and a more consistent interface. 
```

> This shift aligns with the broader industry's move towards [cgroup v2](https://kubernetes.io/docs/concepts/architecture/cgroups/), offering improved functionality, scalability, and a more consistent interface. 

Author: this was addressed by another commit. thanks!

Maintance mode means that no new features will be added to cgroup v1 support. 
Critical security fixes will still be provided, however, bug-fixing is now best-effort, meaning major bugs may be fixed if feasible, but some issues might remain unresolved.

Its recommended that you start switching to use cgroup v2 as soon as possible. For more details on the reasoning behind placing cgroup v1 in maintaince mode and starting the migration process see the [Moving cgroup v1 Support into Maintenance Mode in Kubernetes 1.31](/blog/kubernetes-1-31-moving-cgroup-v1-support-maintenance-mode) article. 

> Please report any problems encounter by filing an [issue](https://github.com/kubernetes/kubernetes).

Reviewer: Even better IMO: https://github.com/kubernetes/kubernetes/issues/new/choose

## Release notes and upgrade actions required

Check out the full details of the Kubernetes 1.31 release in our [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.31.md).

### Scheduler now uses QueueingHint when `SchedulerQueueingHints` is enabled
Added support to the scheduler to start using a QueueingHint registered for Pod/Updated events,
to determine whether updates to  previously unschedulable Pods have made them schedulable.
The new support is active when the feature gate `SchedulerQueueingHints` is enabled.

Previously, when unschedulable Pods were updated, the scheduler always put Pods back to into a queue
(`activeQ` / `backoffQ`). However not all updates to Pods make Pods schedulable, especially considering
many scheduling constraints nowadays are immutable. Under the new behaviour, once unschedulable Pods
are updated, the scheduling queue checks with QueueingHint(s) whether the update may make the
pod(s) schedulable, and requeues them to `activeQ` or `backoffQ` only when at least one
QueueingHint returns `Queue`.

**Action required for custom scheduler plugin developers**: 

> Plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods themselves. Example: suppose you develop a custom plugin that denies Pods that have a `schedulable=false` label. Given Pods with a `schedulable=false` label will be schedulable if the `schedulable=false` label is removed, this plugin would implement QueueingHint for Pod/Update event that returns Queue when such label changes are made in unscheduled Pods. [#122234](https://github.com/kubernetes/kubernetes/pull/122234)

Reviewer 2: ```suggestion
Plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods themselves. Example: suppose you develop a custom plugin that denies Pods that have a `schedulable=false` label. Given Pods with a `schedulable=false` label will be schedulable if the `schedulable=false` label is removed, this plugin would implement QueueingHint for Pod/Update event that returns Queue when such label changes are made in unscheduled Pods. You can find more details in the pull request [#122234](https://github.com/kubernetes/kubernetes/pull/122234).
```

---
layout: blog
title: 'Kubernetes v1.31: TITLE'
date: 2024-08-13
slug: kubernetes-v1-31-release
author: >
  [Kubernetes v1.31 Release Team](https://github.com/kubernetes/sig-release/blob/master/releases/release-1.31/release-team.md)
---

> 

Reviewer: We typically have some text like:
> ```markdown
> Announcing the release of Kubernetes v1.31: <THING>, <tagline>!
> 
> Similar to previous releases, the release of Kubernetes v1.31 introduces new
> stable, beta, and alpha features. <flattering text>. 
> This release consists of <count> enhancements.
> Of those enhancements, <count> have graduated to Stable, <count> are entering Beta, 
> and <count> have graduated to Alpha.
>
> ```

Author: I've updated this. Thanks Tim.

I also place to update our release blog template in the comms handbook b/c its apparently very out of date

or visit the [Service](/docs/concepts/services-networking/service/#traffic-distribution) documentation page.

This work was done as part of [KEP #4444](https://github.com/kubernetes/enhancements/issues/4444) by [SIG Network](https://github.com/kubernetes/community/tree/master/sig-network).

### Kubernetes VolumeAttributesClass ModifyVolume

[VolumeAttributesClass](/docs/concepts/storage/volume-attributes-classes/) API is moving to beta in v1.31.
The VolumeAttributesClass provides a generic,
Kubernetes-native API for modifying dynamically volume parameters like provisioned IO.
This allows workloads to vertically scale their volumes on-line to balance cost and performance, if supported by their provider.
This feature had been alpha since Kubernetes 1.29.

This work was done as a part of [KEP #3751](https://github.com/kubernetes/enhancements/issues/3751) and lead by [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

## New features in Alpha

_This is a selection of some of the improvements that are now alpha following the v1.31 release._
    

> ### New DRA APIs for better accelerators and other hardware management

Reviewer 8: @pohly   can you review this section of the release blog?

Reviewer 9: I can give the LGTM in place of @pohly (who is currently on vacation). This looks good / accurate to me.

Kubernetes v1.31 brings an updated dynamic resource allocation (DRA) API and design. The main focus in the update is on structured parameters because they make resource information and requests transparent to Kubernetes and clients and enable implementing features like cluster autoscaling. DRA support in the kubelet was updated such that version skew between kubelet and the control plane is possible. With structured parameters, the scheduler allocates `ResourceClaims` while scheduling a pod. Allocation by a DRA driver controller is still supported through what is now called "classic DRA". 

With Kubernetes v1.31, classic DRA has a separate feature gate and needs to be enabled with `DRAControlPlaneController`. With such a control plane controller, a DRA driver can implement allocation policies that are not supported yet through structured parameters.

This work was done as part of [KEP #3063](https://github.com/kubernetes/enhancements/issues/3063) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node).
    

> ### Support for volumes based on OCI artifacts

Reviewer 8: @saschagrunert  can you review this section of the release blog?

> ### Support for volumes based on OCI artifacts

Reviewer 10: ```suggestion
### Support for image volumes
```

The Kubernetes community is moving towards fulfilling more Artificial Intelligence (AI) and Machine Learning (ML) use cases in the future. 

> One of the requirements to support these use cases is to support Open Container Initiative (OCI) compatible images and artifacts (referred as OCI objects) directly as a native volume source. This allows users to focus on OCI standards as well as enables them to store and distribute any content using OCI registries.

Reviewer 10: Word duplication "support":
```suggestion
One of the requirements to fulfill these use cases is to support Open Container Initiative (OCI) compatible images and artifacts (referred as OCI objects) directly as a native volume source. This allows users to focus on OCI standards as well as enables them to store and distribute any content using OCI registries.
```

> Given that, the Image Volume Source feature is introduced as a new alpha feature in v1.31. This feature allows users to specify an image reference as volume in a pod while reusing it as volume mount within containers.

Reviewer 10: ```suggestion
Given that, the `ImageVolume` feature is introduced as a new alpha enhancement in v1.31. This feature allows users to specify an image reference as volume in a pod while reusing it as volume mount within containers.
```

Reviewer: I'd write:
```suggestion
Given that, v1.31 adds a new alpha feature to allow using an OCI image as a volume in a Pod.
This feature allows users to specify an image reference as volume in a pod while reusing it as volume
mount within containers. You need to enable the `ImageVolume` feature gate to try this out.
```

This work was done as part of [KEP #4639](https://github.com/kubernetes/enhancements/issues/4639) by [SIG Node](https://github.com/kubernetes/community/tree/master/sig-node) and [SIG Storage](https://github.com/kubernetes/community/tree/master/sig-storage).

> ### Exposing device health information through Pod status

Reviewer 8: @SergeyKanzhelev   can you review this section of the release blog?

Reviewer 11: lgtm
